crates/connector_proxy/src/libs/stream.rs

```diff
@@ -1,25 +1,244 @@
+        let mut buf = BytesMut::with_capacity(4096);
+        match r.read_buf(&mut buf).await {
+            Ok(0) => Ok(None),
+            Ok(_) => Ok(Some((Bytes::from(buf), r))),
+            Err(e) => raise_err(&format!("error during streaming {:?}.", e)),
+        }
+    })
+}
+
+/// Given a stream of bytes, try to deserialize them into Airbyte Messages.
+/// This can be used when reading responses from the Airbyte connector, and will
+/// handle validation of messages as well as handling of AirbyteLogMessages.
+/// Will ignore* messages that cannot be parsed to an AirbyteMessage.
+/// * See https://docs.airbyte.com/understanding-airbyte/airbyte-specification#the-airbyte-protocol
+pub fn stream_airbyte_responses(
+    in_stream: InterceptorStream,
+) -> impl TryStream<Item = std::io::Result<Message>, Ok = Message, Error = std::io::Error> {
+    stream::once(async {
+        let mut buf = BytesMut::new();
+        let items = in_stream
+            .map(move |bytes| {
+                let b = bytes?;
+                buf.extend_from_slice(b.chunk());
+                let chunk = buf.chunk();
+
+                // Deserialize to Value first, instead of Message, to avoid missing 'is_eof' signals in error.
+                let deserializer = Deserializer::from_slice(chunk);
+                let mut value_stream = deserializer.into_iter::<Value>();
+
+                // Turn Values into Messages and validate them
+                let values: Vec<Result<Message, std::io::Error>> = value_stream
+                    .by_ref()
+                    .map_while(|value| match value {
+                        Ok(v) => Some(Ok(v)),
+                        Err(e) => {
+                            // we must stop as soon as we hit EOF to avoid
+                            // progressing value_stream.byte_offset() so that we can
+                            // safely drop the buffer up to byte_offset() and pick up the leftovers
+                            // when working with the next bytes
+                            if e.is_eof() {
+                                return None;
+                            }
+
+                            Some(raise_err(&format!(
+                                "error in decoding JSON: {:?}, {:?}",
+                                e,
+                                std::str::from_utf8(chunk)
+                            )))
+                        }
+                    })
+                    .map(|value| match value {
+                        Ok(v) => {
+                            let message: Message = match serde_json::from_value(v) {
+                                Ok(m) => m,
+                                // We ignore JSONs that are not Airbyte Messages according
+                                // to the specification:
+                                // https://docs.airbyte.com/understanding-airbyte/airbyte-specification#the-airbyte-protocol
+                                Err(_) => return Ok(None),
```

Comment:
Upon a closer read, this isn't right, either. The wording of the spec in ambiguous, and makes it seem like you could expect that each line on stdout is JSON, but that is not so. Here's some completely reasonable assumptions about airbyte connectors that turned out to be wrong:

- stdout will only have JSON: Nope, it'll also contain lines of plain text mixed in
- all logs are emitted as airbyte `Message`s: Nope, see above. The log `Message`s seem to be more for things that are meant to be surfaced directly to the end user, as opposed to debug logs.
- debug logs are written to stderr: Nope, only some of their connectors do that. Many log only to stdout, or to both.

The [behavior we're using currently](https://github.com/estuary/flow/blob/cb92e21981eeaf3d9fcee2f5636c495eb465fe6c/go/capture/driver/airbyte/driver.go#L497) is sort of the opposite of what's here. Lines are allowed to not parse as JSON, but if they _do_ parse as JSON, then they're required to deserialize sucessfully as airbyte messages. Based on yet another read of the spec, we might also need to permit that second case, but it hasn't caused an issue so far.

Reply:
I read the protocol specification once again and my reading still is that each line must be JSON, although it can be a non-AirbyteMessage, in which case it's ignored. https://docs.airbyte.com/understanding-airbyte/airbyte-specification/#the-airbyte-protocol

There may be cases where debug logs are printed to stdout as plaintext ([as in this example](https://github.com/airbytehq/airbyte/pull/11977) which I opened a PR for) but I'm pretty sure it's a bug. However, I think Airbyte _is_ being forgiving about this in their own product which leads to them not finding these bugs out. When I asked one of their engineers about this plaintext print statement, he was very surprised and asked me "I wonder how it's working at all".

Now with that said, we might say we do not trust connectors to conform to the specification properly and we would rather be forgiving to avoid issues on our side, and that's reasonable.

Reply:
Okay, I updated the code so that we now operate on lines, rather on bytes directly. This has simplified the message handling function _a lot_, and now allows us to handle plaintext lines. The serde StreamDeserializer could not handle this (i.e. it would never progress if it sees an input such as: `I am plaintext\n{"x": 2}`) because it does not operate on lines.

Reply:
I am not sure if there is any consequence to switching to operating over bytes of lines. For one thing, I know that each line can be very large, does that cause a problem to have a single large `Bytes` rather than multiple smaller ones? We used to cap our `Bytes` at 4096 bytes.

Reply:
In any case though, I don't see any way at the moment to be able to deserialize lines into JSON while handling plaintext without having to somehow run the deserialize on a line, so in the end we will have to feed "one line" to the deserializer.

Reply:
> so in the end we will have to feed "one line" to the deserializer.

Yeah, I think that's right.

> There may be cases where debug logs are printed to stdout as plaintext (https://github.com/airbytehq/airbyte/pull/11977 which I opened a PR for) but I'm pretty sure it's a bug.

Yeah, I'd agree with that, as well as your suggestion that we should also tolerate it, since that's what airbyte does. Basically, I think we have to treat the "spec" as aspirational, and consider the "real spec" to be whatever airbyte accepts. The rationale is that users will not be very understanding if they can't use an airbyte connector in Flow because of a "bug", when that connector seems to work just fine in Airbyte. I do think that pushing for conformance to the published spec is also important, though, and so ultimately I think we need to take it case by case. The stdout logging issue is just one that we'd seen in (anecdotally) many connectors, but I don't really have a good sense of how common it is. Sorry for the ramble, but I think my conclusion here is just that I'd trust your judgement on what to do here now that you're informed on the issue.